src/sql/src/kafka_util.rs

```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
+            Ok(ts) if ts < 0 => {
+                let now: i64 = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)?
+                    .as_millis()
+                    .try_into()?;
+                let ts = now - ts.abs();
+                if ts <= 0 {
+                    bail!("Relative `time_offset` must be smaller than current system timestamp")
+                }
+                ts
+            }
+            // Timestamp in millis (e.g. 1622659034343)
+            Ok(ts) => ts,
+            _ => bail!("`time_offset` must be a number"),
+        },
+        _ => bail!("`time_offset` must be a number"),
+    };
+
+    // Lookup offsets
+    let mut config = rdkafka::ClientConfig::new();
+    config.set("bootstrap.servers", broker);
+    for (k, v) in client_options {
+        config.set(k, v);
+    }
+
+    let consumer: BaseConsumer<KafkaErrCheckContext> =
+        match config.create_with_context(KafkaErrCheckContext::default()) {
+            Ok(consumer) => consumer,
+            Err(e) => {
+                match e {
+                    rdkafka::error::KafkaError::ClientCreation(s) => {
+                        // Pass existing error back up.
+                        bail!(rdkafka::error::KafkaError::ClientCreation(s))
+                    }
+                    _ => bail!(e),
+                }
+            }
+        };
+
+    task::spawn_blocking({
+        let topic = topic.to_string();
+        move || {
+            // There cannot be more than i32 partitions
+            let num_partitions: i32 = consumer
+                .fetch_metadata(Some(&topic), Duration::from_secs(10))?
+                .topics()
+                .first()
+                .unwrap()
+                .partitions()
+                .len()
+                .try_into()?;
+
+            let mut tpl = TopicPartitionList::with_capacity(1);
+            tpl.add_partition_range(&topic, 0, num_partitions - 1);
+            tpl.set_all_offsets(Offset::Offset(time_offset))?;
+
+            let offsets_for_times = consumer.offsets_for_times(tpl, Duration::from_secs(10))?;
+
+            // Translate to `start_offsets`, filling in the end offset for
+            // partitions which have not reached the provided timestamp yet.
+            let elems = offsets_for_times.elements();
+
+            let mut time_offsets = elems.iter();
+            let mut start_offsets = Vec::with_capacity(num_partitions as usize);
+
+            for pid in 0..num_partitions {
+                let offset = match time_offsets.next() {
+                    Some(partition) => match partition.offset() {
+                        Offset::Offset(offset) => offset,
+                        _ => fetch_end_offset(&consumer, &topic, pid)?,
+                    },
+                    _ => fetch_end_offset(&consumer, &topic, pid)?,
+                };
+                start_offsets.push(offset);
+            }
```

Comment:
In what cases does Kafka return results for fewer partitions than we asked for, is that an expected case we need to gracefully handle? If it's only obscure cases like a partition resize as the query is being purified, I'd be inclined to simplify this and treat that as an invariant violation, eg. (crappy rust pseudocode)

```
let start_offsets = offsets_for_times.iter().map({
    if let Offset::offset(o) {
        o
    } else {
        fetch_end_offset
    }}).collect();

if start_offsets.len() != num_partitions {
  // throw user-facing invariant violation error
}
```

Reply:
The case is independent of partition resizing and occurs if the provided timestamp value is greater than all timestamps for a partition (e.g. skew between partitions or a timestamp far "in the future"). My gut feeling is that it wouldn't be a too common case, but I also think it doesn't hurt to handle it.

I'm not too opinionated here, so feel free to insist on simplifying the code. If we keep it, I'll add a comment to clarify why we do it.

Reply:
This may be me totally misunderstanding the underlying API calls, but I'll take a stab at articulating what I'm concerned about to see if it's legit. 🙂 

My main concern isn't about coming up with end offsets for ts that are out of range, but that iteration is based on num_partitions and we currently match on the None branch of `time_offsets.next()`, so may synthesize offsets for pids not in the response.

---

We create an API request for ListOffsets with a payload like: "pid 0 offset ts FOO, pid 1 offset ts FOO, pid 2 offset ts FOO"

My expectation is that we get back a result for each pid we requested, though the individual partition result may be Offset::Invalid or some form of error to represent FOO was an out of range ts.
eg. response might be: "pid 0 offset 1, pid 1 offset invalid/error, pid 2 offset 3"
We do want to map those invalid offset results to the tail offset of the topic if the ts is out of range, as you mentioned.

What I would be surprised to see is a response with a subset of pids we asked for, such as:
"pid 0 offset 1, pid 1 offset 1"
or worse
"pid 1 offset 1, pid 2 offset 2"

If we never see responses of that form, then the code is correct but the match on None/current iteration style is unnecessary and could hide problems. If we do see responses of that form then the code may be wrong because we're relying on the implicit pid ordering but the response is potentially sparse (we never validate `TopicPartitionListElem.partition()` but instead infer it).

I think my questions are:
1. Are we guaranteed that all pids in our request show up in our response?
2. Are we guaranteed the ordering of pids in the response?

(my hunch is that the answer to both is yes)

Reply:
That's also how I understood your first message. Sorry for being unclear!

The answer to both questions is Yes. Originally, I was under the wrong impression that the answer to 1. is No, because I only checked it against `Offset::Invalid` whereas the response is `Offset::End` (which makes sense).

I'll update the PR.

Reply:
Just pushed a fix in https://github.com/MaterializeInc/materialize/pull/6941/commits/f195b30f7cdaad6d0e866d044a5e2b9461f1f703.

I left it as a separate commit for easier reviewing.